src/components/frontend-engine/use-form-change.ts

```diff
@@ -0,0 +1,154 @@
+	const checkIsFormValid = useCallback(() => {
+		try {
+			hardValidationSchema.validateSync(getValues());
+			return true;
+		} catch (error) {
+			return false;
+		}
+	}, [getValues, hardValidationSchema]);
+
+	const handleValueChange = useRef(() => {
+		// noop
+	});
+	handleValueChange.current = () => {
+		const formValues = getFormValues(undefined, stripUnknown);
+		const formValidity = checkIsFormValid();
+		// attach / fire onValueChange event only when formValidationConfig has values
+		// otherwise isValid will be returned incorrectly as true
+		if (onValueChange && Object.keys(formValidationConfig || {}).length) {
+			// memoise to prevent unnecessary calls
+			if (previousFormValidity.current !== formValidity || !isEqual(previousFormValues.current, formValues)) {
+				onValueChange(formValues, formValidity);
+			}
+		}
+		previousFormValues.current = formValues;
+		previousFormValidity.current = formValidity;
+	};
+
+	// =============================================================================
+	// EFFECTS
+	// =============================================================================
+	useEffect(() => {
+		const subscription = watch((value, { name }) => {
+			if (name) {
+				setField(name, value[name]);
+			} else {
+				setFields(value);
+			}
+		});
+		return () => subscription.unsubscribe();
+	}, []);
+
+	useEffect(() => {
+		// attach / fire onChange event only when formValidationConfig has values
+		// otherwise isValid will be returned incorrectly as true
+		if (onChange && Object.keys(formValidationConfig || {}).length) {
+			const subscription = watch(() => {
+				onChange(getFormValues(undefined, stripUnknown), checkIsFormValid());
+			});
+			onChange(getFormValues(undefined, stripUnknown), checkIsFormValid());
+
+			return () => subscription.unsubscribe();
+		}
+
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [checkIsFormValid, onChange, watch, formValidationConfig]);
+
+	useEffect(() => {
+		const subscription = watch(() => {
+			handleValueChange.current();
+		});
+
+		return () => subscription.unsubscribe();
+	}, [watch]);
+
+	useEffect(() => {
+		// when config changes due to overrides or conditional rendering, mark validity for re-evaluation
+		if (previousFormValidationConfig) {
+			const mapValidationRules = (config: TFormYupConfig) => {
+				return Object.entries(config).map(([fieldId, { validationRules }]) => [fieldId, validationRules]);
+			};
+			const previousValidationRules = mapValidationRules(previousFormValidationConfig);
+			const validationRules = mapValidationRules(formValidationConfig);
+			if (!isEqual(previousValidationRules, validationRules)) {
+				setHasSchemaChange(true);
+			}
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [formValidationConfig]);
+
+	useEffect(() => {
+		// re-evaluate form validity when schema changes
+		if (hasSchemaChange && previousHardValidationSchema !== hardValidationSchema) {
```

Comment:
I vaguely recall you talking about this, can explain again why the `previousHardValidationSchema !== hardValidationSchema` check? since both objects will never match each other

Reply:
in the current behaviour, it just so happens `hasSchemaChange` and `hardValidationSchema` update in the same render cycle, but that may not be guaranteed in the future. it is safer to only trigger this when `hardValidationSchema` has definitely updated